---
title: "Preliminary processing"
output: html_document
date: "2024-03-19"
---

```{r}
library(Seurat)
library(tidyverse)
library(harmony)
library(DoubletFinder)
library(miQC)
library(SoupX)
library(data.table)
library(SeuratWrappers) #For miQC
```

This stage should give 1) processed seurat object for each sample, 2) merged dataset without any QC, 3) whole dataset post-QC, and 4) the preliminary objects for the subsets. Arguably, QC can be done prior to integration, but this workflow applies to datasets that have greater interest in the integrated version, where limited interest is given to single samples.

The collection of stats for the dataset pre-and-post QC is usually needed but can be partially omitted from this Rmd. In general, we want to know 1) the cell number of a sample, dataset, and cluster as well as 2) whether some clusters and samples are particularly low/high in mitochondrial RNA, nUMI, and nFeature detected.

```{r}
source("DoubletFinder_helper.R")
```

# Part 1 important alignment output and organize assays

Note that the codes are not fully optimized for automation

```{r define directories}
indir = "/storage1/fs1/martyomov/Active/collaborations/carisa/Choi_ZBTB46" #Can be made into a command-line input
outdir = "./R_outs/"
```

## 1.1 load assays 
###  load Z46 sample
```{r create treatment object}
sample = "SR003274_Sample1_Z46"
Z46 <- Seurat::Read10X(paste(indir,"/", sample, "/outs/filtered_feature_bc_matrix", sep = ""))

ZBTB46 <- CreateSeuratObject(counts = Z46$`Gene Expression`)
ZBTB46[["ADT"]] <- CreateAssay5Object(counts=Z46$`Antibody Capture`)

# check rownames
rownames(ZBTB46[["ADT"]]) <- c("B0301","B0302") #TME-PE has B0301; TME has B0302
```

### load control
```{r create control object}

sample = "SR003274_Sample2_Control"
ctrl <- Seurat::Read10X(paste(indir,"/", sample, "/outs/filtered_feature_bc_matrix", sep = ""))

Control <- CreateSeuratObject(counts = ctrl$`Gene Expression`)
Control[["ADT"]] <- CreateAssay5Object(counts=ctrl$`Antibody Capture`)
```

## 1.2 Add additional metadata

```{r add QC metrics}
Control[["percent.mt"]] = PercentageFeatureSet(object = Control, pattern = "^MT-|^mt-",assay = "RNA")
Control$nFeature_RNA_log10 <- log10(Control$nFeature_RNA)
Control$nCount_RNA_log10 <- log10(Control$nCount_RNA)
Control$nCount_ADT_log10 <- log10(Control$nCount_ADT)
Control$B0301 <- Control[["ADT"]]$counts[1,]
Control$B0302 <- Control[["ADT"]]$counts[2,]
Control$B0301_log10 <- log10(Control$B0301 + 1)
Control$B0302_log10 <- log10(Control$B0302 + 1)

ZBTB46[["percent.mt"]] = PercentageFeatureSet(object = ZBTB46, pattern = "^MT-|^mt-",assay = "RNA")
ZBTB46$nFeature_RNA_log10 <- log10(ZBTB46$nFeature_RNA)
ZBTB46$nCount_RNA_log10 <- log10(ZBTB46$nCount_RNA)
ZBTB46$nCount_ADT_log10 <- log10(ZBTB46$nCount_ADT)
ZBTB46$B0301 <- ZBTB46[["ADT"]]$counts[1,]
ZBTB46$B0302 <- ZBTB46[["ADT"]]$counts[2,]
ZBTB46$B0301_log10 <- log10(ZBTB46$B0301 + 1)
ZBTB46$B0302_log10 <- log10(ZBTB46$B0302 + 1)
```

```{r add idents}
Control$orig.ident <- rep("Control",ncol(Control[["RNA"]]))
ZBTB46$orig.ident <- rep("ZBTB46",ncol(Control[["RNA"]]))
Idents(Control) <- Control$orig.ident
Idents(ZBTB46) <- ZBTB46$orig.ident
```

```{r make plots}
FeatureScatter(Control, feature1 = "nFeature_RNA_log10", feature2 = "percent.mt")+
  ggtitle("Control")
FeatureScatter(Control, feature1 = "nFeature_RNA", feature2 = "percent.mt")+
  ggtitle("Control")
FeatureScatter(ZBTB46, feature1 = "nFeature_RNA_log10", feature2 = "percent.mt")+
  ggtitle("ZBTB46")
FeatureScatter(ZBTB46, feature1 = "nFeature_RNA", feature2 = "percent.mt")+
  ggtitle("ZBTB46")

VlnPlot(Control,features = c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"),assay = "RNA", layer = "counts")
```

## 1.3 perform miQC for each sample

### Control

```{r}
Control <- RunMiQC(Control, percent.mt = "percent.mt", nFeature_RNA = "nFeature_RNA", posterior.cutoff = 0.75,
               model.slot = "flexmix_model")
#PlotMiQC(Control, color.by = "miQC.probability",model.slot = "flexmix_model") + ggplot2::scale_color_gradient(low = "grey", high = "purple")+ggtitle("Control")
#PlotMiQC(Control, color.by = "miQC.keep")+ggtitle("Control")

table(Control$miQC.keep)

#hist(Control$miQC.probability)
```

```{r}
VlnPlot(Control,features = c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"), split.by = "miQC.keep",split.plot = T,assay = "RNA", layer = "counts")
miQC_kept <- subset(Control, miQC.keep == "keep")
VlnPlot(miQC_kept,features = c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"),assay = "RNA", layer = "counts")
```

### Z46


```{r}
ZBTB46 <- RunMiQC(ZBTB46, percent.mt = "percent.mt", nFeature_RNA = "nFeature_RNA", posterior.cutoff = 0.75,
               model.slot = "flexmix_model")
#PlotMiQC(ZBTB46, color.by = "miQC.probability") + ggplot2::scale_color_gradient(low = "grey", high = "purple")+ggtitle("ZBTB46")
#PlotMiQC(ZBTB46, color.by = "miQC.keep")+ggtitle("ZBTB46")


table(ZBTB46$miQC.keep)
```


```{r}
VlnPlot(ZBTB46,features = c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"), split.by = "miQC.keep",split.plot = T,assay = "RNA", layer = "counts")
miQC_kept <- subset(ZBTB46, miQC.keep == "keep")
VlnPlot(miQC_kept,features = c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"),assay = "RNA", layer = "counts")
```

## 1.4 Process single sample and run DoubletFinder

### Control

```{r}
Control <- NormalizeData(object = Control,assay = "RNA", normalization.method = "LogNormalize", scale.factor = 10000)
Control <- FindVariableFeatures(object = Control, selection.method = 'mean.var.plot', mean.cutoff = c(0.0125, 5), dispersion.cutoff = c(0.5, Inf))
VariableFeatures(Control) <-  VariableFeatures(Control)[!grepl("^TRA|^TRB|^IGH|^IGK|MAMU-A", VariableFeatures(Control))]
VariableFeaturePlot(Control)

Control <- ScaleData(object = Control, assay = "RNA", features = VariableFeatures(object = Control), vars.to.regress = c("nCount_RNA", "percent.mt"))


Control<- RunPCA(Control,features =  VariableFeatures(object = Control),assay = "RNA")


ElbowPlot(Control,ndims = 40)

Control <- RunUMAP(Control, dims = 1:20)
Control <- FindNeighbors(object =Control, dims = 1:20)              
Control <- FindClusters(object = Control, resolution = 0.1)

DimPlot(Control,reduction = "umap")
FeaturePlot(Control,features = c("percent.mt","nFeature_RNA_log10"),reduction = "umap")
FeaturePlot(Control,features = c("percent.mt","nCount_RNA_log10"),reduction = "umap")

```


```{r add ashtag as metadata }
table(Control$nFeature_ADT)
table(ZBTB46$nFeature_ADT)

Control$B0301_raw <- Control[["ADT"]]$counts[1,]
Control$B0302_raw <- Control[["ADT"]]$counts[2,]
```


```{r run on control}
#no ground truth 
sweep.res.list <- paramSweep(Control, PCs = 1:30, sct = FALSE)
sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)
bcmvn <- find.pK(sweep.stats)


homotypic.prop <- modelHomotypic(Control@meta.data$seurat_clusters)          
nExp_poi <- round(0.075*nrow(Control@meta.data))  ## Assuming 7.5% doublet formation rate 
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))

Control <- doubletFinder(Control, PCs = 1:30, pN = 0.25, pK = 0.09, nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)

head(Control@meta.data)
# you may keep only the predictions; note that DoubletFinder is not working well
```

```{r save object}
save(Control, file = paste0(outdir,"Control_full.Robj"))
```

### treatment

```{r}
ZBTB46 <- NormalizeData(object = ZBTB46,assay = "RNA", normalization.method = "LogNormalize", scale.factor = 10000)
ZBTB46 <- FindVariableFeatures(object = ZBTB46 , selection.method = 'mean.var.plot', mean.cutoff = c(0.0125, 5), dispersion.cutoff = c(0.5, Inf))
VariableFeatures(ZBTB46) <-  VariableFeatures(ZBTB46)[!grepl("^TRA|^TRB|^IGH|^IGK|MAMU-A", VariableFeatures(ZBTB46))]
VariableFeaturePlot(ZBTB46)

ZBTB46 <- ScaleData(object = ZBTB46, assay = "RNA", features = VariableFeatures(object = ZBTB46), vars.to.regress = c("nCount_RNA", "percent.mt"))
ZBTB46 <- RunPCA(ZBTB46,features =  VariableFeatures(object = ZBTB46),assay = "RNA")

ElbowPlot(ZBTB46,ndims = 40)

ZBTB46 <- RunUMAP(ZBTB46, dims = 1:30)
ZBTB46 <- FindNeighbors(object = ZBTB46, dims = 1:30)              
ZBTB46 <- FindClusters(object = ZBTB46, resolution = 0.1)

DimPlot(ZBTB46,reduction = "umap")
FeaturePlot(ZBTB46,features = c("percent.mt","nFeature_RNA_log10"),reduction = "umap")
FeaturePlot(ZBTB46,features = c("percent.mt","nCount_RNA_log10"),reduction = "umap")
```

```{r}
ZBTB46$B0301 <- ZBTB46[["ADT"]]$scale.data[1,]
ZBTB46$B0302 <- ZBTB46[["ADT"]]$scale.data[2,]
```

```{r}
sweep.res.list <- paramSweep(ZBTB46, PCs = 1:30, sct = FALSE)
sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)

bcmvn <- find.pK(sweep.stats)
homotypic.prop <- modelHomotypic(ZBTB46@meta.data$seurat_clusters)          
nExp_poi <- round(0.075*nrow(ZBTB46@meta.data))  ## Assuming 7.5% doublet formation rate 
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))

ZBTB46 <- doubletFinder(ZBTB46, PCs = 1:30, pN = 0.25, pK = 0.09, nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)
```

```{r}
save(ZBTB46, file = paste0(outdir,"ZBTB46_full.Robj"))
```

# Part 2 merge preliminary dataset for record keeping

## 2.1 merge and create stat checking visualization

```{r}
whole <- merge(x=Control,y=ZBTB46,add.cell.ids=c("Control","ZBTB46"))

ctrl_range = range(grep("^Control",rownames(whole@meta.data)))

whole@meta.data$orig.ident[ctrl_range[1]:ctrl_range[2]] <- "Control"

Z46_range = range(grep("^ZBTB46",rownames(whole@meta.data)))

whole@meta.data$orig.ident[Z46_range[1]:Z46_range[2]] <- "ZBTB46"

table(whole@meta.data$orig.ident)
```

```{r }
save(whole,file = paste0(outdir,"whole_full.Robj"))
```

```{r}
VlnPlot(whole, features="percent.mt",split.by = "orig.ident")
VlnPlot(whole, features="nFeature_RNA_log10",split.by = "orig.ident")
VlnPlot(whole, features="nCount_RNA_log10",split.by = "orig.ident")
VlnPlot(whole, features=c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"),group.by = "orig.ident")
VlnPlot(whole,features = c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"), split.by = "miQC.keep",group.by = "orig.ident")
```


```{r}
ggplot(whole@meta.data)+
  geom_histogram(aes(x=percent.mt,y = after_stat(width*density*100)),bins = 100)+
  facet_wrap(vars(orig.ident),ncol=1)+
  geom_vline(xintercept = 5,col="red") +
  ylab("proportion")+
  xlab("percentage of mitochondrial RNA")
```

```{r}
ggplot(whole@meta.data, aes(x = log10(nFeature_RNA))) +
  geom_histogram(bins = 100, aes(y = after_stat(width*density*100))) +
  theme_classic() +
  facet_wrap(~orig.ident, ncol = 1) +
  geom_vline(xintercept = 2.5,col="red") + 
  geom_vline(xintercept = 4,col="red") + 
  ggtitle("nGene") +
  ylab("")
```


```{r}
ggplot(whole@meta.data, aes(x = log10(nCount_RNA))) +
  geom_histogram(bins = 100, aes(y = after_stat(width*density*100))) +
  theme_classic() +
  facet_wrap(~orig.ident, ncol = 1) +
  geom_vline(xintercept = 2.5,col="red") + 
  geom_vline(xintercept = 4.5,col="red") + 
  ggtitle("nCount") +
  ylab("")
```

```{r}
ggplot(whole@meta.data, aes(x = log10(nCount_ADT))) +
  geom_histogram(bins = 100, aes(y = after_stat(width*density*100))) +
  theme_classic() +
  facet_wrap(~orig.ident, ncol = 1) +
  geom_vline(xintercept = 2,col="red") + 
  geom_vline(xintercept = 4.5,col="red") + 
  ggtitle("nCount") +
  ylab("")
```

```{r}
ggplot(whole@meta.data, aes(x = nFeature_RNA)) +
  geom_histogram(bins = 100, aes(y = after_stat(width*density*100))) +
  theme_classic() +
  facet_wrap(~orig.ident, ncol = 1) +
  geom_vline(xintercept = 300,col="red") + 
  geom_vline(xintercept = 5000,col="red") + 
  ggtitle("nGene") +
  ylab("")
```

## 2.2 process and integrate

```{r}
whole <- NormalizeData(object = whole, normalization.method = "LogNormalize", scale.factor = 10000)
#there is no need to normalize again; normalization is cell specific, but you can still run it.
whole <- FindVariableFeatures(object = whole, selection.method = 'mean.var.plot', mean.cutoff = c(0.0125, 5), dispersion.cutoff = c(0.5, Inf))
VariableFeatures(whole) <-  VariableFeatures(whole) [!grepl("^Tra|^Trb|^Igh|^Igk|Mamu-A", VariableFeatures(whole))]
# Human: VariableFeatures(whole) <-  VariableFeatures(whole) [!grepl("^TRA|^TRB|^IGH|^IGK|MAMU-A", VariableFeatures(whole))]
# you may also consider VariableFeatures(whole) <-  VariableFeatures(whole) [!grepl("^Gm|*\Rik$", VariableFeatures(whole))]

VariableFeaturePlot(whole)

whole <- ScaleData(object = whole, features = VariableFeatures(object = whole), vars.to.regress = c("nCount_RNA", "percent.mt"))
whole <- RunPCA(object = whole,
                features =  VariableFeatures(object = whole),
                dims = 1:50)
gc()
ElbowPlot(whole,ndims = 50) # Elbow plot is needed for picking nPCs
# You can choose to check the correlation between specific metadata and the PCs, but it is more informative when there are sampling constraints and you have multiple co-variates to correct for.
```


```{r}
whole <- RunHarmony(object = whole, group.by.vars = c("orig.ident"), max.iter.harmony = 20)
whole <- RunUMAP(whole,dims = 1:40,reduction = "harmony")
whole <- RunTSNE(whole,dims = 1:40,reduction = "harmony")

whole <- FindNeighbors(whole, dims = 1:40,reduction = "harmony")
for(res in seq(0.1, 1, 0.1))  {   
  whole <- FindClusters(object = whole, resolution = res, print.output = 0, save.SNN = TRUE)
} 
gc()
save(whole,file = paste0(outdir,"whole_full_wHarmony.Robj"))
```

## 2.3 visualization of metrics

```{r}
VlnPlot(whole, features="percent.mt",split.by = "orig.ident")
VlnPlot(whole, features="nFeature_RNA_log10",split.by = "orig.ident")
VlnPlot(whole, features="nCount_RNA_log10",split.by = "orig.ident")
VlnPlot(whole, features=c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"),group.by = "orig.ident")
VlnPlot(whole,features = c("percent.mt","nFeature_RNA_log10","nCount_RNA_log10"), split.by = "miQC.keep",group.by = "orig.ident")
```


```{r}

FeaturePlot(whole, features = "nFeature_RNA",reduction = "umap", min.cutoff =3000 )
FeaturePlot(whole, features = "nCount_RNA",reduction = "umap",min.cutoff = 10000)

RidgePlot(whole,features = c("nCount_RNA","nFeature_RNA"))+RotatedAxis()
```

```{r}
table(whole$RNA_snn_res.1)
```

## 2.4 apply preliminary QC threshold

```{r}
table(whole$miQC.keep)
table(whole$DF_classification)
whole <- subset(whole,subset = miQC.keep == "keep")
whole <- subset(whole,subset = DF_classification == "Singlet")
nrow(whole@meta.data)
```

```{r}
save(whole,file = paste0(outdir,"whole_filtered_by_pred.Robj"))
```

```{r}
whole <- FindVariableFeatures(object = whole, selection.method = 'mean.var.plot', mean.cutoff = c(0.0125, 5), dispersion.cutoff = c(0.5, Inf))
VariableFeatures(whole) <-  VariableFeatures(whole) [!grepl("^TRA|^TRB|^IGH|^IGK|MAMU-A", VariableFeatures(whole))]
VariableFeaturePlot(whole)

whole <- ScaleData(object = whole, features = VariableFeatures(object = whole), vars.to.regress = c("nCount_RNA", "percent.mt"))
whole <- RunPCA(object = whole,
                features =  VariableFeatures(object = whole),
                dims = 1:50)
gc()
ElbowPlot(whole,ndims = 50)

whole <- RunHarmony(object = whole, group.by.vars = c("orig.ident"), max.iter.harmony = 20)
whole <- RunUMAP(whole,dims = 1:30,reduction = "harmony")
whole <- RunTSNE(whole,dims = 1:30,reduction = "harmony")

whole <- FindNeighbors(whole, dims = 1:30,reduction = "harmony")
for(res in seq(0.1, 1, 0.1))  {   
  whole <- FindClusters(object = whole, resolution = res, print.output = 0, save.SNN = TRUE)
} 
gc()
whole <- JoinLayers(whole)
save(whole,file = paste0(outdir,"whole_filtered_by_pred.Robj"))
```

## 2.5 determine hashtag labels

```{r}
g <- ggplot(whole@meta.data,aes(x=log10(B0301+1),y=log10(B0302+1)))+
  stat_density_2d(aes(fill = after_stat(level)), geom = "polygon", colour="white")+
  theme_bw()
ggsave(filename = paste0(outdir,"hashtag_raw_counts_2D_density.pdf"), plot = g, units = "cm", height = 30,width = 30 )
```

```{r}
g <- ggplot(whole@meta.data,aes(x=log10(B0301+1),y=log10(B0302+1)))+
  stat_density_2d(aes(fill = after_stat(level)), geom = "polygon", colour="white")+
  facet_wrap(~orig.ident)+
  theme_bw()
ggsave(filename = paste0(outdir,"hashtag_raw_counts_2D_density_by_sample.pdf"), plot = g, units = "cm", height = 30,width = 30 )
```


```{r}
g <- ggplot(whole@meta.data,aes(x=log10(B0301+1),y=log10(B0302+1),color = orig.ident))+
  geom_point() +
  theme_bw()
ggsave(filename = paste0(outdir,"hashtag_raw_counts_scatter.pdf"), plot = g, units = "cm", height = 30,width = 30 )
```

```{r}
whole@meta.data <- whole@meta.data %>%
  mutate(hashtag = case_when(
  (B0301_log10 > 2.5 ) & (B0302_log10 > 2.5) ~ "double_high",
  (B0301_log10 > 2.5 ) & (B0302_log10 <= 2.5) ~ "Reporter+",
  (B0301_log10 <= 2.5) & (B0302_log10 > 2.5) ~ "Reporter-",
  .default = "double_low"
))
```

```{r}
ggplot(whole@meta.data) +
 geom_histogram(aes(x = log10(B0301+1), fill = "B0301"), alpha = 0.5,binwidth = 0.1) +
 geom_histogram(aes(x = log10(B0302+1), fill = "B0302"), alpha = 0.5,binwidth = 0.1) +
 scale_fill_manual(values = c("B0301" = "red", "B0302" = "blue")) +
 geom_vline(xintercept = 2.5,col = "purple")+
 facet_wrap(~orig.ident,scales = "free",dir = "v")+
 labs(title = "Overlaid log10 Hashtag Counts", x = "log10(counts+1)", y = "Frequency")+
  guides(fill=guide_legend(title="Hashtag"))
```


```{r}
save(whole,file = paste0(outdir,"whole_filtered_by_pred.Robj"))
```

## 2.6 Check markers and create subsets for second round of cleaning


```{r}
DimPlot(whole, reduction = "umap",group.by = "orig.ident",shuffle = T)
DimPlot(whole, reduction = "umap",group.by = "RNA_snn_res.0.1")
DimPlot(whole, reduction = "umap")

DimPlot(whole, reduction = "umap",group.by = "RNA_snn_res.0.3")

DotPlot(whole,features = c("Cd3e","Cd8a","Cd4","Cd19","Adgre1","Ncam1","Xcr1","Ncr1","Siglech","Mki67","S100a8","Fcer1a","Fcer2a","Flt3","Pecam1","Myct1"),group.by = "RNA_snn_res.0.1")+RotatedAxis()

DotPlot(whole,features = c("Cd3e","Cd8a","Cd4","Cd19","Adgre1","Ncam1","Xcr1","Ncr1","Siglech","Mki67","S100a8","Fcer1a","Fcer2a","Flt3","Pecam1","Myct1"))+RotatedAxis()
```

```{r}
Idents(whole) <- whole$RNA_snn_res.0.1
lymphoid <- subset(whole, idents = c(1,5))
myeloid <- subset(whole,idents = c(0,2,3,6,7))
stromal <- subset(whole, idents = c(4,8))
```

```{r}
save(lymphoid,file = paste0(outdir,"lymphoid.Robj"))
save(myeloid,file = paste0(outdir,"myeloid.Robj"))
save(stromal,file = paste0(outdir,"stromal.Robj"))
```


