---
title: "Pub2_conventional_T_and_clonality"
output: html_document
date: "2024-08-04"
---

```{r}
library(Seurat)
library(tidyverse)
library(RColorBrewer)
```

```{r}
pal <- c(brewer.pal(n = 12,name = "Set3"),brewer.pal(n = 12,name = "Paired"),brewer.pal(n = 8,name = "Accent"))
pal[2] <- "darkred"
pal[9] <- "grey25"
pal[23] <- "turquoise4"
pal[19] <- colors()[60]
pal[13] <- "darkblue"
pal[28] <- colors()[16]
pal[31] <- colors()[54] 
pal <- c(pal,colors()[c(85,84,83,89,45,31,41,42,26,29,10,139,107,108,120,109,119,121,143)])

```

```{r}
RDSdir <- "../V2_processing_analysis/R_outs/"
outdir <- "plots/"
if (!dir.exists(outdir)) dir.create(outdir)
```

# general plots

```{r}
load(paste0(RDSdir,"lymphoid_convT_V3p2_merged_scRep.Robj"))
```

## conventional T cell UMAP with the subsets

### update annotation

```{r}
DimPlot(screp,group.by = "RNA_snn_res.0.9",label = T)
```

```{r}
screp@meta.data <- screp@meta.data %>% mutate(
  annotation_final = case_match(
    as.character(RNA_snn_res.0.9),
    "0" ~ "CD8_Tcf7High",
    "1" ~ "CD8_ToxHigh_Tcf7Low",
    "2" ~ "CD4_Th1",
    "3" ~ "NKT",
    "4" ~ "CD8_IFN",
    "5" ~ "CD8_ItgaeHigh",
    "6" ~ "CD8_PD1Low",
    "7" ~ "CD8_IL18rapHigh",
    "8" ~ "Ccr7High",
    "9" ~ "Mki67+",
    "10" ~ "CD4_IFN",
    "11" ~ "CD4_Treg",
    "12" ~ "CD4_Th2",
    "13" ~ "CD8_Bcl3"
  )
)
```

### UMAP plot
```{r}
p1 <- DimPlot(screp.aa,group.by = "annotation_final",label = T,pt.size = 1,label.box = T,label.color = "white",repel = T)+scale_color_manual(values = pal)+scale_fill_manual(values = pal)+guides(color=guide_legend(ncol = 3, position = "bottom",override.aes = list(size=5)))
p2 <- DimPlot(screp.aa,group.by = "annotation_final",label = F,pt.size = 1,label.box = F,label.color = "white",repel = T)+scale_color_manual(values = pal)+scale_fill_manual(values = pal)+guides(color=guide_legend(ncol = 3, position = "bottom",override.aes = list(size=5)))

g <- gridExtra::grid.arrange(grobs = list(p1,p2),ncol = 2)
ggsave(paste0(outdir,"/Fig4_conventional_T_cell_UMAP.pdf"),plot = g, width = 15,height = 8)
```

### Marker Heatmap

```{r}
markers <- read.table("SR003099_DF_V3p1_convT_scRep_irf8d32tapoe_resolution_0.9_markers.tsv",sep = "\t",header = T)
```

```{r}
markers <- markers %>% mutate(
  annotation_final = as.factor(case_match(
    as.character(cluster),
    "0" ~ "CD8_Tcf7High",
    "1" ~ "CD8_ToxHigh_Tcf7Low",
    "2" ~ "CD4_Th1",
    "3" ~ "NKT",
    "4" ~ "CD8_IFN",
    "5" ~ "CD8_ItgaeHigh",
    "6" ~ "CD8_PD1Low",
    "7" ~ "CD8_IL18rapHigh",
    "8" ~ "Ccr7High",
    "9" ~ "Mki67+",
    "10" ~ "CD4_IFN",
    "11" ~ "CD4_Treg",
    "12" ~ "CD4_Th2",
    "13" ~ "CD8_Bcl3"
  )
))
```

```{r}
top_markers <- markers %>% filter(!(gene %in% markers$gene[grep("^Gm|Rik$", markers$gene)])) %>% group_by(annotation_final) %>% filter(pct.1 > 0.75) %>% slice_max(order_by = c(avg_log2FC), n = 20) %>% ungroup() 
```

```{r}
g <- DoHeatmap(screp.aa, features = top_markers$gene,group.by = "annotation_final",draw.lines = F,size = 5,group.colors = pal,raster = FALSE,disp.max = 1.5,disp.min = -1.5)
g
ggsave(filename = paste0(outdir,"/Fig4_marker_heatmap.pdf"),width = 18,height = 18)
```

### density plot

```{r}
screp.aa@meta.data <- screp.aa@meta.data %>% mutate(
  celltype = as.factor(case_match(
    annotation_final,
    "CD8_Tcf7High" ~ "CD8T",
    "CD8_ToxHigh_Tcf7Low"  ~ "CD8T",
    "CD4_Th1" ~ "CD4T",
    "NKT" ~ "NKT",
    "CD8_IFN" ~ "CD8T",
    "CD8_ItgaeHigh" ~ "CD8T",
    "CD8_PD1Low" ~ "CD8T" ,
    "CD8_IL18rapHigh" ~ "CD8T",
    "Mki67+" ~ "Proliferating T",
    "CD4_IFN" ~ "CD4T",
    "CD4_Treg" ~ "CD4T",
    "CD4_Th2" ~ "CD4T",
    "CD8_Bcl3" ~ "CD8T",
    .default = celltype
  )
))
```

```{r}
save(screp.aa, file = paste0(RDSdir,"lymphoid_convT_V3p2_merged_scRep.Robj"))
```

```{r}
meta <- cbind(screp.aa@meta.data,screp.aa@reductions$umap@cell.embeddings)
```


```{r}
g <- ggplot(meta, aes(x = umap_1, y = umap_2)) +
  geom_point(aes(color = annotation_final,fill = annotation_final),alpha = 0.8,size = 0.5,fill = "white") +
  geom_density_2d_filled(alpha = 0.5) +
  geom_density_2d(colour = "black",alpha = 0.6)+
  scale_fill_manual(values = c('white','#fed976','#feb24c','#fd8d3c','#fc4e2a','#e31a1c','#b10026'))+ #'#ffffd4'
  scale_color_manual(values = pal)+
  theme_classic()+
  xlim(-6.5,11)+
  ylim(-8,6)+
  facet_wrap(~orig.ident,ncol = 2, scales = "free")+
 guides(fill = guide_legend(title = "Level",ncol = 2,position = "bottom"),
        color = guide_legend(ncol = 3,position = "bottom",title = "Annotation",override.aes = list(alpha = 0,shape = "circle")))
g
ggsave(filename = paste0(outdir,"/Fig4_density_plot.pdf"),width = 12,height = 10)
```

```{r}
g <- ggplot(meta, aes(x = umap_1, y = umap_2)) +
  geom_point(aes(color = annotation_final,fill = annotation_final),alpha = 0.8,size = 0.5,fill = "white") +
  geom_density_2d(colour = "black",alpha = 0.6)+
  scale_color_manual(values = pal)+
  theme_classic()+
  xlim(-6.5,11)+
  ylim(-8,6)+
  facet_wrap(~orig.ident,ncol = 2, scales = "free")
g
ggsave(filename = paste0(outdir,"/Fig4_density_plot_without_filling.pdf"),plot = g,width = 12,height = 10)
```

```{r}
# calculate within cell type proportions
table(meta$celltype)
```



```{r}
CD4_prop <- meta %>% filter(celltype == "CD4T") %>% group_by(orig.ident,annotation_final) %>% summarise(n_subpop = n()) %>%
  ungroup() %>% group_by(orig.ident) %>% mutate(subpop_prop = n_subpop *100/sum(n_subpop))
```

```{r}
table(CD4_prop$annotation_final)
```

```{r}
ggplot(CD4_prop) +
  geom_bar(aes(x = orig.ident,y = subpop_prop, fill = annotation_final), stat = "identity",position = "stack")+
  scale_fill_manual(values = pal[1:5])+theme_classic()
```

```{r}
CD8_prop <- meta %>% filter(celltype == "CD8T") %>% group_by(orig.ident,annotation_final) %>% summarise(n_subpop = n()) %>%
  ungroup() %>% group_by(orig.ident) %>% mutate(subpop_prop = n_subpop *100/sum(n_subpop))
table(CD8_prop$annotation_final)
```
```{r}
ggplot(CD8_prop) +
  geom_bar(aes(x = orig.ident,y = subpop_prop, fill = annotation_final), stat = "identity",position = "stack")+
  scale_fill_manual(values = pal[c(1,6:12)])+theme_classic()
```

```{r}
p1 <- ggplot(CD4_prop) +
  geom_bar(aes(x = orig.ident,y = subpop_prop, fill = annotation_final), stat = "identity",position = "stack")+
  scale_fill_manual("Subpopulation",values = pal[1:5])+theme_classic()+xlab("Phenotype")+ylab("% in CD4")
p2 <- ggplot(CD8_prop) +
  geom_bar(aes(x = orig.ident,y = subpop_prop, fill = annotation_final), stat = "identity",position = "stack")+
  scale_fill_manual("Subpopulation",values = pal[c(1,6:12)])+theme_classic()+xlab("Phenotype")+ylab("% in CD8")
g <- gridExtra::grid.arrange(grobs = list(p1,p2), ncol = 2)
ggsave(filename = paste0(outdir,"/Fig4_subtype_proportion_bar.pdf"),plot = g,width = 10,height = 6)
```

## DE

### four condition comparison

Let's create a psuedobulk by condition

```{r}
bulk <- AggregateExpression(screp,group.by = "orig.ident",return.seurat = T)
```


```{r}
NT_C_bulk <- subset(bulk,subset = orig.ident == "NT-C")
NT_KO_bulk <- subset(bulk,subset = orig.ident == "NT-KO")
T_C_bulk <- subset(bulk,subset = orig.ident == "T-C")
T_KO_bulk <- subset(bulk,subset = orig.ident == "T-KO")
```

```{r}
hist(NT_KO_bulk@assays$RNA@layers$data)
```


```{r}
NT_changes = log2((NT_KO_bulk@assays$RNA@layers$data+1)/(NT_C_bulk@assays$RNA@layers$data+1))
T_changes = log2((T_KO_bulk@assays$RNA@layers$data+1)/(T_C_bulk@assays$RNA@layers$data+1))
```

```{r}
hist(NT_changes)
```


```{r}
qua_df <- data.frame(NT_fc = NT_changes, T_fc = T_changes,genes = rownames(bulk)) %>% 
  mutate(labels = case_when(
    NT_fc < 1 & NT_fc > -1 & T_fc < 1 & T_fc > -1 ~ "constant",
    .default = "variable"

  ))
```

```{r}
ggplot(qua_df)+
  geom_point(aes(x = NT_fc, y = NT_fc,color = labels))+
  geom_hline(yintercept = c(-1,1))+
  geom_vline(xintercept = c(-1,1))+
  xlab("Average log2fc NT-KO/NT-C")+
  ylab("Average log2fc T-KO/T-C")
```

Direct psuedobulk doesn't work

### cross-condition DE
Let's try directly DE; 

T-C v.s. NT-C
T-KO v.s. T-C *use the avg log2 FC value here for four qua plot
NT-KO v.s. NT-C *use the avg log2 FC value here for four qua plot

```{r}
de_clusters <- function(current_cluster) {
  groups <- gsub(" ","_",paste0(conditions,"_",current_cluster))
  both_markers <- FindMarkers(screp, ident.1=groups[1], ident.2=groups[2], test.use="wilcox", min.pct=0.00, logfc.threshold = 0) # change name of object 
  marker_genes <- row.names(both_markers)
  mutate(both_markers, gene = marker_genes, cluster = current_cluster)
}
```

#### T-C v.s. NT-C 

```{r}
screp$group <- gsub(" ","_",paste0(screp$orig.ident,"_",screp$annotation_final))
conditions <- unique(screp$orig.ident)[order(unique(screp$orig.ident))][c(3,1)] # Levels: NT_C NT_KO T_C T_KO
clusters <- unique(screp$annotation_final)
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(screp) <- screp$group
```

```{r}
table(screp$group)
```


```{r}

for (i in clusters) {
  output_file <- paste0("",paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0("",paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  volcano_plot_file <- paste0(outdir,paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '_volcano.pdf'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  g <- ggplot(de_results_tbl,aes(x=avg_log2FC, y = -log10(p_val_adj), label = gene))+
  geom_point(aes(color = threshold))+
  geom_vline(xintercept = c(1,-1),color = "red2")+geom_hline(yintercept = 1.3, color = "red2")+ ggrepel::geom_label_repel(aes(label=ifelse((p_val_adj < 0.05) & (abs(avg_log2FC) > 1),as.character(gene),'')),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  size = 2,
                  segment.color = 'grey50',
                  max.overlaps = 30) +
  theme_classic()+
    ggtitle(paste0(i,": NT-C v.s. T-C"))
  ggsave(volcano_plot_file,plot = g, width = 6, height = 6)
  
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}
```


#### NT-KO v.s. NT-C

```{r}
screp$group <- gsub(" ","_",paste0(screp$orig.ident,"_",screp$annotation_final))
conditions <- unique(screp$orig.ident)[order(unique(screp$orig.ident))][1:2] # Levels: NT_C NT_KO T_C T_KO
clusters <- unique(screp$annotation_final)
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(screp) <- screp$group
```

```{r}
for (i in clusters) {
  output_file <- paste0("",paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0("",paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  volcano_plot_file <- paste0(outdir,paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '_volcano.pdf'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  g <- ggplot(de_results_tbl,aes(x=avg_log2FC, y = -log10(p_val_adj), label = gene))+
  geom_point(aes(color = threshold))+
  geom_vline(xintercept = c(1,-1),color = "red2")+geom_hline(yintercept = 1.3, color = "red2")+ ggrepel::geom_label_repel(aes(label=ifelse((p_val_adj < 0.05) & (abs(avg_log2FC) > 1),as.character(gene),'')),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  size = 2,
                  segment.color = 'grey50',
                  max.overlaps = 30) +
  theme_classic()+
    ggtitle(paste0(i,": NT-KO v.s. NT-C"))
  ggsave(volcano_plot_file,plot = g, width = 6, height = 6)
  
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}
```


#### T-KO v.s. T-C

```{r}
screp$group <- gsub(" ","_",paste0(screp$orig.ident,"_",screp$annotation_final))
conditions <- unique(screp$orig.ident)[order(unique(screp$orig.ident))][3:4] # Levels: NT_C NT_KO T_C T_KO
clusters <- unique(screp$annotation_final)
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(screp) <- screp$group
```

```{r}
for (i in clusters) {
  output_file <- paste0("",paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0("",paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  volcano_plot_file <- paste0(outdir,paste0("convT_",conditions[1],'_vs_',conditions[2],'_', i, '_volcano.pdf'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  g <- ggplot(de_results_tbl,aes(x=avg_log2FC, y = -log10(p_val_adj), label = gene))+
  geom_point(aes(color = threshold))+
  geom_vline(xintercept = c(1,-1),color = "red2")+geom_hline(yintercept = 1.3, color = "red2")+ ggrepel::geom_label_repel(aes(label=ifelse((p_val_adj < 0.05) & (abs(avg_log2FC) > 1),as.character(gene),'')),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  size = 2,
                  segment.color = 'grey50',
                  max.overlaps = 30) +
  theme_classic()+
    ggtitle(paste0(i,": T-KO v.s. T-C"))
  ggsave(volcano_plot_file,plot = g, width = 6, height = 6)
  
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}
```

### VlnPlot for some important genes

```{r}
genes <- c("Ccr7","Lef1","Tcf7","Ccr2","Gzma","Gzmb","Gzmk","Prf1","Tigit","Tox","Pdcd1","Icos","Ifit2","Il2ra","Il6","Pdcd1lg2")
```

```{r}
screp.aa$group <- gsub(" ","_",paste0(screp.aa$orig.ident,"_",screp.aa$celltype))
```


```{r}
DotPlot(screp.aa,features = genes, group.by = "group")+RotatedAxis()
```

```{r}
VlnPlot(screp.aa,features = "Gzmk",group.by = "celltype",split.by = "orig.ident",alpha = 0.2)
```

```{r}
glist <- lapply(genes, function(x){
  VlnPlot(screp.aa,features = x,group.by = "celltype",split.by = "orig.ident",alpha = 0.2)
})
g <- gridExtra::grid.arrange(grobs = glist, ncol = 3)
ggsave(filename = paste0(outdir,"/Fig4_T_cell_VlnPlots_group_by_celltype.pdf"),plot = g,width = 20,height = 20)

glist <- lapply(genes, function(x){
  VlnPlot(screp.aa,features = x,group.by = "orig.ident",alpha = 0.2)
})
g <- gridExtra::grid.arrange(grobs = glist, ncol = 3)
ggsave(filename = paste0(outdir,"/Fig4_T_cell_VlnPlots_group_by_genotype.pdf"),plot = g,width = 20,height = 20)
```


## clonality analysis

require re-do; performed in a different Rmd

```{r}
colnames(screp.aa@meta.data)
```

screp plots were made separately
### Frequency plot

```{r}
table(meta$cloneSize)
```


```{r}
g <- ggplot(meta, aes(x = umap_1, y = umap_2)) +
  geom_point(aes(color = factor(as.character(cloneSize),levels = c("None ( < X <= 0)",
                                                      "Rare (0 < X <= 1e-04)",
                                                      "Small (1e-04 < X <= 0.001)",
                                                      "Medium (0.001 < X <= 0.01)",
                                                      "Large (0.01 < X <= 0.1)",
                                                      "Hyperexpanded (0.1 < X <= 1)"))),alpha = 1,size = 0.5,fill = "white") +
  geom_density_2d_filled(alpha = 0.2) +
  geom_density_2d(colour = "black",alpha = 0.6)+
  scale_fill_manual(values = c('white','#fed976','#feb24c','#fd8d3c','#fc4e2a','#e31a1c','#b10026'))+ #'#ffffd4'
  scale_color_manual(values = c("None ( < X <= 0)" = "white",
                                "Rare (0 < X <= 1e-04)" = "beige",
                                "Small (1e-04 < X <= 0.001)" = "orange",
                                "Medium (0.001 < X <= 0.01)" = "coral",
                                "Large (0.01 < X <= 0.1)" = "red2",
                                "Hyperexpanded (0.1 < X <= 1)" = "darkred"),na.value = "grey95")+
  theme_classic()+
  facet_wrap(~orig.ident,ncol = 2, scales = "free")+
 guides(fill = guide_legend(title = "Level",ncol = 2,position = "bottom"),
        color = guide_legend(ncol = 3,position = "bottom",title = "Clone Freuqnecy",override.aes = list(alpha = 0,shape = "circle")))+
  xlim(-6.5,11)+
  ylim(-8,6)

g
ggsave(filename = paste0(outdir,"/Fig4_cloneType_plot.pdf"),width = 12,height = 10)
```


```{r}
g <- ggplot(meta, aes(x = umap_1, y = umap_2)) +
  geom_point(aes(color = factor(as.character(cloneSize),levels = c("None ( < X <= 0)",
                                                      "Rare (0 < X <= 1e-04)",
                                                      "Small (1e-04 < X <= 0.001)",
                                                      "Medium (0.001 < X <= 0.01)",
                                                      "Large (0.01 < X <= 0.1)",
                                                      "Hyperexpanded (0.1 < X <= 1)"))),alpha = 1,size = 0.5,fill = "white") +
  geom_density_2d_filled(alpha = 0.2) +
  geom_density_2d(colour = "black",alpha = 0.6)+
  scale_fill_manual(values = c('white','#fed976','#feb24c','#fd8d3c','#fc4e2a','#e31a1c','#b10026'))+ #'#ffffd4'
  scale_color_manual(values = c("None ( < X <= 0)" = "white",
                                "Rare (0 < X <= 1e-04)" = "beige",
                                "Small (1e-04 < X <= 0.001)" = "orange",
                                "Medium (0.001 < X <= 0.01)" = "coral",
                                "Large (0.01 < X <= 0.1)" = "red2",
                                "Hyperexpanded (0.1 < X <= 1)" = "darkred"),na.value = "grey95")+
  theme_classic()+
  facet_wrap(~orig.ident,ncol = 2, scales = "free")+
 guides(fill = guide_legend(title = "Level",ncol = 2,position = "bottom"),
        color = guide_legend(ncol = 3,position = "bottom",title = "Clone Freuqnecy",override.aes = list(alpha = 0,shape = "circle")))+
  xlim(-6.5,11)+
  ylim(-8,6)

g
ggsave(filename = paste0(outdir,"/Fig4_cloneType_plot_without_filling.pdf"),width = 12,height = 10)
```

### Gini Index

```{r}
library(DescTools)
```

```{r}
Gini_df <- meta %>% filter((!is.na(cloneSize))) %>% group_by(orig.ident,celltype) %>% mutate(Gini_index = Gini(clonalFrequency)) %>% ungroup()
```

```{r}
table(Gini_df$Gini_index)
```


```{r}
ggplot(Gini_df) +
  geom_bar(aes(x = orig.ident, y = Gini_index, fill = orig.ident),stat = "identity",position = "dodge")+
  facet_wrap(~celltype)
```

Trouble with replicating the Gini index plot using scRep results...Try updating cell type

```{r}
DimPlot(screp.aa,group.by = "celltype")
DimPlot(screp.aa,group.by = "annotation_final")
```


```{r}
Gini_df  <- screp.aa@meta.data %>% filter((!is.na(cloneSize))) %>% group_by(orig.ident,celltype,CTaa,clonalFrequency) %>% summarise(cloneFreq = n()) %>% ungroup() %>% group_by(orig.ident,celltype) %>% mutate(Gini_index = Gini(clonalFrequency)) %>% ungroup()
```



```{r}
g <- ggplot(Gini_df) +
  geom_bar(aes(x = orig.ident, y = Gini_index, fill = orig.ident),stat = "identity",position = "dodge")+
  facet_wrap(~celltype,scales = "free")+xlab("Condition")+theme_classic()+ylab("Gini Index")
g
ggsave(filename = paste0(outdir,"/Fig4_Gini_index_plot.pdf"),width = 10,height = 6)
```


### barplot of clone sizes

```{r}
freq_meta <- screp.aa@meta.data %>% filter(!is.na(cloneSize)) %>% group_by(orig.ident,cloneSize) %>% mutate(type_counts = n()) %>% ungroup() %>% group_by(orig.ident) %>% mutate(sample_size = n()) %>% ungroup() %>% mutate(proportion = type_counts/sample_size)
```

```{r}
table(freq_meta$cloneSize)
```

```{r}
# this plot is a replicate from scRep
g <- ggplot(freq_meta) + 
  geom_bar(aes(x = orig.ident, y = proportion, fill = cloneSize),stat = "summary",position = "stack")
g
ggsave(paste0(outdir,"Fig4_stacked_cloneType.pdf"),plot = g, width = 10, height = 10)
```

```{r}
freq_meta %>% group_by(orig.ident, CTaa) %>% summarise(cloneFreq = n()) %>% slice_max(order_by = cloneFreq, n = 20) %>% ungroup()
```

```{r}
freq_meta %>% group_by(orig.ident, CTaa,clonalFrequency) %>% summarise() %>% ungroup() %>% group_by(orig.ident) %>% slice_max(order_by = clonalFrequency, n = 20) %>% ungroup()%>% arrange(clonalFrequency)
```

```{r}
g <- ggplot(freq_meta %>% group_by(orig.ident,celltype,CTaa,clonalFrequency) %>% summarise() %>% ungroup() %>% group_by(orig.ident) %>% slice_max(order_by = clonalFrequency, n = 20) %>% ungroup()) + 
  geom_bar(aes(x = reorder(CTaa, -clonalFrequency), y = clonalFrequency, fill = orig.ident),stat = "identity",position = "dodge")+
  facet_grid(~celltype, scale = "free")+theme_classic()+
  xlab("Clonotype by amino acid")+
  theme(axis.text.x = element_text(angle = 90,size = 5))
g
ggsave(paste0(outdir,"SAMPLE_Fig4_CD3r_chain_aa_top20.pdf"),plot = g, width = 20, height = 20)
```



```{r}
write.table(freq_meta %>% group_by(orig.ident, CTaa,clonalFrequency) %>% summarise() %>% ungroup() %>% group_by(orig.ident) %>% slice_max(order_by = clonalFrequency, n = 20) %>% ungroup()%>% arrange(clonalFrequency),
            file = "screp_top20_CTaa_by_sample.tsv",sep = "\t",col.names = T,row.names = F)
```

```{r}
CD3r.table <- freq_meta %>% group_by(orig.ident, CTgene,Frequency) %>% summarise() %>% ungroup() %>% group_by(orig.ident) %>% mutate(proportion = Frequency/sum(Frequency)) %>%  slice_max(order_by = proportion, n = 5) 
CD3r.table$CTgene <- paste0(CD3r.table$orig.ident,":",CD3r.table$CTgene)
```

```{r}
g <- ggplot(CD3r.table)+
  geom_bar(aes(x = orig.ident, y = proportion, fill = CTgene),stat = "identity",position = "stack")+
  theme(legend.text = element_text(size = 5),
        legend.title = element_text(size = 8))+
  scale_fill_manual(values = pal)+
  xlab("Genotype")+
  guides(fill = guide_legend(ncol = 4,position = "bottom"))
g
ggsave(paste0(outdir,"SAMPLE_Fig4_CD3r_chain_gene_top5.pdf"),plot = g, width = 15, height = 10)
```

### more visualization of top clones

```{r}
top_AA <- freq_meta %>% group_by(orig.ident, CTaa,clonalFrequency) %>% summarise() %>% ungroup() %>% group_by(orig.ident) %>% slice_max(order_by = clonalFrequency, n = 20,with_ties = F) %>% ungroup()%>% arrange(clonalFrequency)
```

```{r}
table(top_AA$clonalFrequency)
```

```{r}
meta <- meta %>% 
  mutate(clone_type = case_when(
    CTaa %in% top_AA$CTaa ~ "Top",
    .default = "non-Top"
  ))
```

```{r}
g <- ggplot(meta)+
  geom_point(aes(x= umap_1, y = umap_2, color = clone_type,size = clonalProportion))+
  facet_wrap(~orig.ident)+scale_color_manual(values = c("grey50","red2"))+scale_radius(range = c(0,2))+theme_classic()
g
ggsave(paste0(outdir,"Fig4_top_clones_on_umap_by_geno.pdf"),plot = g, width = 10, height = 8)
```

```{r}
top_prolif_AA <- freq_meta %>% group_by(orig.ident, CTaa,clonalFrequency,celltype) %>% summarise() %>% ungroup() %>% group_by(orig.ident) %>% slice_max(order_by = clonalFrequency, n = 20,with_ties = F) %>% ungroup()%>% arrange(clonalFrequency) %>% filter(celltype == "Proliferating T")
```

```{r}
top_prolif_AA$CTaa
```
```{r}
meta <- meta %>% mutate(
  prolif_clone = case_when(
    !(CTaa %in% top_prolif_AA$CTaa) ~ "Other",
    .default = CTaa
  ),
  prolif_highlight = case_when(
    !(CTaa %in% top_prolif_AA$CTaa) ~ "Other",
    .default = "Prolif"
  )
)
```

```{r}
table(meta$prolif_clone)
```

```{r}
g <- ggplot(meta)+
  geom_point(aes(x= umap_1, y = umap_2, color = prolif_clone,size = factor(prolif_highlight,levels = c("Other","Prolif")),alpha = factor(prolif_highlight,levels = c("Other","Prolif"))))+scale_size_manual(values = c("Other" = 0.5,"Prolif" = 1))+
  facet_wrap(~orig.ident)+scale_color_manual("Clonotype",values = c(pal[1:7],pal[9:10],pal[13:16],pal[18],"grey70"))+scale_alpha_manual(values = c("Other" = 0.1,"Prolif" = 1))+theme_classic() + guides(alpha = "none",size = "none",colour = guide_legend(override.aes = list(size=3)))
g

ggsave(paste0(outdir,"Fig4_prolif_clones_on_umap.pdf"),plot = g, width = 12, height = 8)
```



# creating phantasus object for conventional T cell pseudobulk

```{r}
library(Biobase)
library(dplyr)
## averaging adta

expr.avg <- AverageExpression(screp, group.by = c("orig.ident"))
expr.avg <- round(expr.avg$RNA, 4)
gene.names <- rownames(expr.avg)

fData <- data.frame(gene = expr.avg@Dimnames[[1]])

rownames(fData) <- fData$gene
## do annotaiton that will be displayed in phantasus
## your annotation can be different from the example

pData <- data.frame(#cluster =  gsub(pattern = "g(.+)_(.+)", replacement = "\\1", x = expr.avg@Dimnames[[2]]),
                    orig.ident =  gsub(pattern = "g(.+)_(.+)", replacement = "\\1", x = expr.avg@Dimnames[[2]]))
rownames(pData) <- expr.avg@Dimnames[[2]]


metadata <- data.frame(labelDescription=names(pData),
                       row.names=names(pData))
phenoData <- new("AnnotatedDataFrame", data=pData, varMetadata=metadata)

metadata <- data.frame(labelDescription=c("Gene_name"),
                       row.names=c("gene"))
featureData <- new("AnnotatedDataFrame", data=fData, varMetadata=metadata)


    ed <- new ("MIAME",
              name="SR003099",
              lab="",
              title="pseudobulk of conventional T cells subset grouped by genotype condition(TE4+Irf8d32)",
              contact="",
              pubMedIds="",
              url="http://artyomovlab.wustl.edu/scn-m/?token=SR003099_DF_V3p1_convT_scRep_irf8d32tapoe",
              other=list())

## create object
eset <- ExpressionSet(assayData = as.matrix(expr.avg), phenoData = phenoData, featureData = featureData, experimentData = ed)
ess <- list(convT_geno_bulk=eset)


save(ess, file=file.path("./", "SR003099_DF_V3p1_convT_bulk_irf8d32tapoe.rda"), version = 2)

```



```{r}
mat <- as.data.frame(as.matrix(expr.avg))
mat <- cbind(rownames(mat),mat)
colnames(mat)[1] <- "gene"
```


```{r}
write.table(mat,file = "convT_V3p1_bulk_exp.tsv",sep = "\t",row.names = F,col.names = T,quote = F)
write.table(pData,file = "convT_V3p1_bulk_col_ann.tsv",sep = "\t",row.names = T,col.names = T,quote = F)
write.table(fData,file = "convT_V3p1_bulk_row_ann.tsv",sep = "\t",row.names = T,col.names = T,quote = F)
```

