---
title: "4-25_differential_expression"
output: html_document
date: "2024-07-13"
---

Previously, DE and GSEA were performed in one notebook. This time I'll perform them separately.

```{r}
# For DE
library(tidyverse)
library(Seurat)
library(harmony)
library(MAST)
```


```{r}
library(RColorBrewer)
library(ggrepel)
pal <- c(brewer.pal(n = 12,name = "Set3"),brewer.pal(n = 12,name = "Paired"),brewer.pal(n = 8,name = "Accent"))
pal[2] <- "darkred"
pal[9] <- "grey25"
pal[23] <- "turquoise4"
pal[19] <- colors()[60]
pal[13] <- "darkblue"
pal[28] <- colors()[16]
pal[31] <- colors()[54] 
pal <- c(pal,colors()[c(85,84,83,89,45,31,41,42,26,29,10,139,107,108,120,109,119,121,143)])
```

# check cluster proportions for deciding what to do for clusters

## Glioma

```{r}
load("data/GBM_RNA_10_glioma_V3_alt.Robj")
```

```{r}
colnames(Glioma@meta.data)
```

```{r}
DimPlot(Glioma, group.by = "annotation_V1")+scale_color_manual(values = pal)
```


```{r}
percentage <- Glioma@meta.data %>% group_by(annotation_V1,progress,batch_ID) %>% summarise(n = n()) %>% ungroup() %>% group_by(progress,batch_ID) %>%
  mutate(percentage = n / sum(n)) %>% ungroup %>% arrange(percentage) %>%
  mutate(labels = scales::percent(percentage))
```

```{r}
g <- ggplot(percentage, aes(x = "", y = percentage, fill = annotation_V1)) +
  geom_col(color = "black") +
  geom_label_repel(aes(label = labels),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE,max.overlaps = 15) +
  guides(fill = guide_legend(title = "Main Annotation",ncol = 2)) +
  scale_fill_manual(values = pal) +
  facet_wrap(~progress+batch_ID) +
  coord_polar(theta = "y") +
  theme_void()
g
ggsave("Glioma_V3_ann_V1_proportion_pie_split_by_batch_progress.pdf",plot = g, width = 15, height = 15)
```
```{r}
percentage <- Glioma@meta.data %>% group_by(annotation_V1,progress) %>% summarise(n = n()) %>% ungroup() %>% group_by(progress) %>%
  mutate(percentage = n / sum(n)) %>% ungroup %>% arrange(percentage) %>%
  mutate(labels = scales::percent(percentage))
```


```{r}
g <- ggplot(percentage, aes(x = "", y = percentage, fill = annotation_V1)) +
  geom_col(color = "black") +
  geom_label_repel(aes(label = labels),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE,max.overlaps = 15) +
  guides(fill = guide_legend(title = "Main Annotation",ncol = 2)) +
  scale_fill_manual(values = pal) +
  facet_wrap(~progress) +
  coord_polar(theta = "y") +
  theme_void()
g
ggsave("Glioma_V3_ann_V1_proportion_pie.pdf",plot = g, width = 15, height = 9)
```
Only tumor 3 and 10 are obviously expanded significantly in HGG


## MG

```{r}
load("data/GBM_RNA_10_microglia_V3p1_alt.Robj")
```

```{r}
percentage <- microglia@meta.data %>% group_by(annotation_V2,progress) %>% summarise(n = n()) %>% ungroup() %>% group_by(progress) %>%
  mutate(percentage = n / sum(n)) %>% ungroup %>% arrange(percentage) %>%
  mutate(labels = scales::percent(percentage))
```

```{r}
g <- ggplot(percentage, aes(x = "", y = percentage, fill = annotation_V2)) +
  geom_col(color = "black") +
  geom_label_repel(aes(label = labels),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE,max.overlaps = 15) +
  guides(fill = guide_legend(title = "Main Annotation",ncol = 2)) +
  scale_fill_manual(values = pal) +
  facet_wrap(~progress) +
  coord_polar(theta = "y") +
  theme_void()
g
ggsave("microglia_V3_ann_V1_proportion_pie.pdf",plot = g, width = 15, height = 9)
```

## lymphoids

```{r}
load("data/GBM_RNA_10_lymphoid_V3p1_alt.Robj")
```

```{r}
percentage <- lymphoid@meta.data %>% group_by(annotation_V2,progress) %>% summarise(n = n()) %>% ungroup() %>% group_by(progress) %>%
  mutate(percentage = n / sum(n)) %>% ungroup %>% arrange(percentage) %>%
  mutate(labels = scales::percent(percentage))
```

```{r}
g <- ggplot(percentage, aes(x = "", y = percentage, fill = annotation_V2)) +
  geom_col(color = "black") +
  geom_label_repel(aes(label = labels),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE,max.overlaps = 15) +
  guides(fill = guide_legend(title = "Main Annotation",ncol = 2)) +
  scale_fill_manual(values = pal) +
  facet_wrap(~progress) +
  coord_polar(theta = "y") +
  theme_void()
g
ggsave("lymphoid_V3_ann_V1_proportion_pie.pdf",plot = g, width = 15, height = 9)
```

## myeloid

```{r}
load("data/GBM_RNA_10_myeloid_V3p2_alt.Robj")
```

```{r}
percentage <- myeloid@meta.data %>% group_by(annotation_V2,progress) %>% summarise(n = n()) %>% ungroup() %>% group_by(progress) %>%
  mutate(percentage = n / sum(n)) %>% ungroup %>% arrange(percentage) %>%
  mutate(labels = scales::percent(percentage))
```

```{r}
g <- ggplot(percentage, aes(x = "", y = percentage, fill = annotation_V2)) +
  geom_col(color = "black") +
  geom_label_repel(aes(label = labels),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE,max.overlaps = 15) +
  guides(fill = guide_legend(title = "Main Annotation",ncol = 2)) +
  scale_fill_manual(values = pal) +
  facet_wrap(~progress) +
  coord_polar(theta = "y") +
  theme_void()
g
ggsave("myeloid_V3_ann_V1_proportion_pie.pdf",plot = g, width = 15, height = 9)
```

## VS subset

```{r}
load("data/GBM_RNA_10_vascular_V3_alt.Robj")
```

```{r}
percentage <- vascular@meta.data %>% group_by(annotation_V2,progress) %>% summarise(n = n()) %>% ungroup() %>% group_by(progress) %>%
  mutate(percentage = n / sum(n)) %>% ungroup %>% arrange(percentage) %>%
  mutate(labels = scales::percent(percentage))
```

```{r}
g <- ggplot(percentage, aes(x = "", y = percentage, fill = annotation_V2)) +
  geom_col(color = "black") +
  geom_label_repel(aes(label = labels),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE,max.overlaps = 15) +
  guides(fill = guide_legend(title = "Main Annotation",ncol = 2)) +
  scale_fill_manual(values = pal) +
  facet_wrap(~progress) +
  coord_polar(theta = "y") +
  theme_void()
g
ggsave("VS_V3_ann_V1_proportion_pie.pdf",plot = g, width = 15, height = 9)
```
catergorize EC and fibroblasts subtypes; 

## NG subset

```{r}
load("data/GBM_RNA_10_resident_V3_alt.Robj")
```

```{r}
percentage <- resident@meta.data %>% group_by(annotation_V2,progress) %>% summarise(n = n()) %>% ungroup() %>% group_by(progress) %>%
  mutate(percentage = n / sum(n)) %>% ungroup %>% arrange(percentage) %>%
  mutate(labels = scales::percent(percentage))
```

```{r}
g <- ggplot(percentage, aes(x = "", y = percentage, fill = annotation_V2)) +
  geom_col(color = "black") +
  geom_label_repel(aes(label = labels),
            position = position_stack(vjust = 0.5),
            show.legend = FALSE,max.overlaps = 15) +
  guides(fill = guide_legend(title = "Main Annotation",ncol = 2)) +
  scale_fill_manual(values = pal) +
  facet_wrap(~progress) +
  coord_polar(theta = "y") +
  theme_void()
g
ggsave("NG_V3_ann_V1_proportion_pie.pdf",plot = g, width = 15, height = 9)
```


# DE for across clusters (subpopulation characterization)

```{r}
indir <- "data/"
outdir <- "DE_analysis_post_meeting2/"
dir.create(outdir)
```

## Glioma

```{r}
Idents(Glioma) <- "annotation_V1"
glioma_DE <- FindAllMarkers(Glioma,test.use = "MAST")
```

```{r}
write.table(glioma_DE,file = paste0(outdir,"glioma_ann_V1.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```


## MG
from 3-55:

NT-high: 0 (Trem2-Ccr6-high), 4 (Fobs_Dusp7), 9 Sparc+ BAM-like
Heterogeneous: 5 (unkown MG)
LGG: 1 (Nr4a1_Fos high),2 (Il7r_Hspa1a),7 (Ifn_resp)
HGG: 6 (PAM-like), 8 (proliferating), 3 (Chst2_high)

```{r}
Idents(microglia) <- "annotation_V2"
MG_DE <- FindAllMarkers(microglia,test.use =  "MAST")
```

```{r}
write.table(MG_DE,file = paste0(outdir,"microglia_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

## macrophages



```{r}
table(myeloid$annotation_main)
```


```{r}
Mo_MPh <- subset(myeloid, subset = annotation_main %in% c("BAM","GAM","Macrophage","Mono-mac","Monocytes"))
```

```{r}
table(Mo_MPh$annotation_V2)
```

```{r}
Idents(Mo_MPh) <- "annotation_V2"
MoMph_DE <- FindAllMarkers(Mo_MPh,test.use =  "MAST")
```

```{r}
write.table(MoMph_DE,file = paste0(outdir,"Mo_MPh_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

## lymphoid

### B cells

```{r}
table(lymphoid$annotation_main)
```


```{r}
B_cells <- subset(lymphoid,subset = annotation_main == "B_cells")
```


```{r}
Idents(B_cells) <- "annotation_V2"
B_DE <- FindAllMarkers(B_cells,test.use =  "MAST")
```

```{r}
write.table(B_DE,file = paste0(outdir,"B_cells_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

### T cells

```{r}
T_cells <- subset(lymphoid, subset =  annotation_main %in% c("CD4_T_cells","CD8_T_cells","T_cells"))
```

```{r}
Idents(T_cells) <- "annotation_V2"
T_DE <- FindAllMarkers(T_cells,test.use =  "MAST")
```

```{r}
write.table(T_DE,file = paste0(outdir,"T_cells_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

## stromal-vascular

### overall by major cell types (note that annotation main is from V1)

```{r}
table(vascular$annotation_V2)
```

```{r}
vascular@meta.data <- vascular@meta.data %>%
  mutate(annotation_main = case_match(
    annotation_V2,
    "Arteriole_EC" ~ "EC",
    "Capillary_EC_1" ~ "EC",
    "Capillary_EC_2" ~ "EC",
    "Capillary_EC_3" ~ "EC",
    "Fibroblast_1" ~ "Fibroblast",
    "Fibroblast_2" ~ "Fibroblast",
    "Fibroblast_3_CAF" ~ "Fibroblast",
    "Pericytes" ~ "Perivascular",
    "Smooth_muscle" ~ "Perivascular",
    .default = annotation_V2
  ))
```

```{r}
table(vascular$annotation_main)
```

```{r}
save(vascular,file = "data/GBM_RNA_10_vascular_V3_alt.Robj")
```


```{r}
Idents(vascular) <- "annotation_main"
VS_main_DE <- FindAllMarkers(vascular,test.use =  "MAST")
```


```{r}
write.table(VS_main_DE,file = paste0(outdir,"VS_ann_main_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

### EC subpopulations

```{r}
EC <- subset(vascular, subset = annotation_main == "EC")
```

```{r}
table(EC$annotation_V2)
```


```{r}
Idents(EC) <- "annotation_V2"
EC_DE <- FindAllMarkers(EC,test.use =  "MAST")
```

```{r}
write.table(EC_DE,file = paste0(outdir,"EC_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```


### Fibroblast subpopulations

```{r}
Fib <- subset(vascular, subset = annotation_main == "Fibroblast")
Idents(Fib) <- "annotation_V2"
Fib_DE <- FindAllMarkers(Fib,test.use =  "MAST")
```

```{r}
write.table(Fib_DE,file = paste0(outdir,"Fib_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```



## neuronal-glial

### Overall cell types

```{r}
table(resident$annotation_V2)
table(resident$annotation_main)
```

```{r}
resident@meta.data <- resident@meta.data %>%
  mutate(annotation_main = case_match(
    annotation_V2,
    "Astrocyte_1" ~ "Astrocyte",
    "Astrocyte_2" ~ "Astrocyte",
    "Astrocyte_3" ~ "Astrocyte",
    "Astrocyte_4" ~ "Astrocyte",
    "Astrocyte_5" ~ "Astrocyte",
    "Astrocyte_6" ~ "Astrocyte",
    "Astrocyte_7" ~ "Astrocyte",
    "Ependymal_1" ~ "Ependymal",
    "Ependymal_2" ~ "Ependymal",
    "Neuron_1" ~ "Neuron",
    "Neuron_2" ~ "Neuron",
    "Oligodendrocyte_1" ~ "Oligodendrocyte",
    "Oligodendrocyte_2" ~ "Oligodendrocyte",
    "Oligodendrocyte_3" ~ "Oligodendrocyte",
    "Progenitor-like_1" ~ "Progenitor-like",
    "Progenitor-like_2" ~ "Progenitor-like",
    .default = annotation_V2
  ))
```

```{r}
table(resident$annotation_main)
```

```{r}
save(resident,file = "data/GBM_RNA_10_resident_V3_alt.Robj")
```

```{r}
resident <- JoinLayers(resident)
```

```{r}
Idents(resident) <- "annotation_main"
NG_main_DE <- FindAllMarkers(resident,test.use =  "MAST")
write.table(NG_main_DE,file = paste0(outdir,"NG_ann_main_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

```{r}
resident@assays[["RNA"]]@cells@.Data <- resident@assays[["RNA"]]@cells@.Data[,c("scale.data.1","counts","data","scale.data")]
```

```{r}
resident@assays[["RNA"]]@features@.Data <- resident@assays[["RNA"]]@features@.Data[,c("scale.data.1","counts","data","scale.data")]
```



### OLG

```{r}
OLG <- subset(resident, subset = annotation_main == "Oligodendrocyte")
table(OLG$annotation_V2)
```

```{r}
Idents(OLG) <- "annotation_V2"
DimPlot(OLG,label = T)
```


```{r}
OLG_DE <- FindAllMarkers(OLG,test.use =  "MAST")
write.table(OLG_DE,file = paste0(outdir,"OLG_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

### Astrocytes

```{r}
AST <- subset(resident, subset = annotation_main == "Astrocyte")
Idents(AST) <- "annotation_V2"
AST_DE <- FindAllMarkers(AST,test.use =  "MAST")

write.table(AST_DE,file = paste0(outdir,"AST_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```


### Ependymal

```{r}
EPD <- subset(resident, subset = annotation_main == "Ependymal")
Idents(EPD) <- "annotation_V2"
EPD_DE <- FindAllMarkers(EPD,test.use =  "MAST")

write.table(EPD_DE,file = paste0(outdir,"EPD_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

### progenitor-like


```{r}
Prog <- subset(resident, subset = annotation_main == "Progenitor-like")
Idents(Prog) <- "annotation_V2"
Prog_DE <- FindAllMarkers(Prog,test.use =  "MAST")

write.table(Prog_DE,file = paste0(outdir,"Prog_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```


### Neuron

```{r}
Neu <- subset(resident, subset = annotation_main == "Neuron")
Idents(Neu) <- "annotation_V2"
Neu_DE <- FindAllMarkers(Neu,test.use =  "MAST")

write.table(Neu_DE,file = paste0(outdir,"Neu_ann_V2.tsv"),sep = "\t",col.names = T,row.names = F,quote = F)
```

# DE across conditions


## Glioma

### Cross-condition DE

```{r}
de_clusters <- function(current_cluster) {
  groups <- gsub(" ","_",paste0(conditions,"_",current_cluster))
  both_markers <- FindMarkers(Glioma, ident.1=groups[1], ident.2=groups[2], test.use="MAST", min.pct=0.00, logfc.threshold = 0) # change name of object 
  marker_genes <- row.names(both_markers)
  mutate(both_markers, gene = marker_genes, cluster = current_cluster)
}
```

```{r}
table(Glioma$annotation_V1)
```

#### LGG v.s. NT;

Tumor_4 exist in some exist in NT, I want to check if these are differentially expressed. The hypothesis is that cells in cluster tumor 4 from NT samples may be normal cells with some resemblance to tumor cells, or the other way around.

```{r}
Glioma$group <- gsub(" ","_",paste0(Glioma$progress,"_",Glioma$annotation_V1))
conditions <- unique(Glioma$progress)[order(unique(Glioma$progress))][2:3]
clusters <- "tumor_4"
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(Glioma) <- Glioma$group
```



```{r}
table(Glioma$group)
```


```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("Glioma_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("Glioma_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}

 
```


#### HGG v.s. LGG

```{r}
Glioma <- subset(Glioma, subset = progress != "No tumor")
```

```{r}
Glioma$group <- gsub(" ","_",paste0(Glioma$progress,"_",Glioma$annotation_V1))
conditions <- unique(Glioma$progress)[order(unique(Glioma$progress))][1:2]
clusters <- paste0("tumor_",c(1,11,2,4,5,6,7,8,9))
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(Glioma) <- Glioma$group
```

```{r}
table(Glioma$group)
```


```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("Glioma_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("Glioma_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}
```


## Microglia

HGG and LGG: unknown_MG, Trem2_Ccr6, Fosb_Dusp1, Chst2_high, Ifn_responsive, Il7r, Nr4a1
LGG and NT:  unknown_MG, Trem2_Ccr6, Fosb_Dusp1, Chst2_high

```{r}
de_clusters <- function(current_cluster) {
  groups <- gsub(" ","_",paste0(conditions,"_",current_cluster))
  both_markers <- FindMarkers(microglia, ident.1=groups[1], ident.2=groups[2], test.use="MAST", min.pct=0.00, logfc.threshold = 0) # change name of object 
  marker_genes <- row.names(both_markers)
  mutate(both_markers, gene = marker_genes, cluster = current_cluster)
}
```

```{r}
table(microglia$annotation_V2)
```

### LGG and NT

```{r}
microglia$group <- gsub(" ","_",paste0(microglia$progress,"_",microglia$annotation_V2))
conditions <- unique(microglia$progress)[order(unique(microglia$progress))][2:3]
clusters <- c("HET_unknown_MG","NT-HIGH_Trem2_Ccr6_high_MG","NT-HIGH_Fosb_Dusp1_high_MG","HGG-HIGH_Chst2_high_MG")
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(microglia) <- microglia$group
```


```{r}
table(microglia$group)
```

```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("MG_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("MG_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}

```

### HGG and LGG


```{r}
microglia$group <- gsub(" ","_",paste0(microglia$progress,"_",microglia$annotation_V2))
conditions <- unique(microglia$progress)[order(unique(microglia$progress))][1:2]
clusters <- c("HET_unknown_MG","NT-HIGH_Trem2_Ccr6_high_MG","NT-HIGH_Fosb_Dusp1_high_MG","HGG-HIGH_Chst2_high_MG","LGG-HIGH_Ifn-responsive_MG","LGG-HIGH_Il7r_Hspa1a_high_MG","LGG-HIGH_Nr4a1_Fos_high_MG")
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(microglia) <- microglia$group
```


```{r}
table(microglia$group)
```

```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("MG_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("MG_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}

```



## lymphoid

```{r}
de_clusters <- function(current_cluster) {
  groups <- gsub(" ","_",paste0(conditions,"_",current_cluster))
  both_markers <- FindMarkers(lymphoid, ident.1=groups[1], ident.2=groups[2], test.use="MAST", min.pct=0.00, logfc.threshold = 0) # change name of object 
  marker_genes <- row.names(both_markers)
  mutate(both_markers, gene = marker_genes, cluster = current_cluster)
}
```

```{r}
table(lymphoid$annotation_V2)
```


LGG and HGG: 0_activated_CD8, 1_activated_CD4, Memory_B_1, Memory_B_2, Memory_B_3, NKT, NK, Naive_T_cells
LGG and NT: Pdcd1_Cd44_high_activated_T, gdT_cells, NKT, NK, Naive_T_cells

### NT and LGG

```{r}
lymphoid$group <- gsub(" ","_",paste0(lymphoid$progress,"_",lymphoid$annotation_V2))
conditions <- unique(lymphoid$progress)[order(unique(lymphoid$progress))][2:3]
clusters <- c("Pdcd1_Cd44_high_activated_T", "gdT_cells", "NKT", "NK", "Naive_T_cells")
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(lymphoid) <- lymphoid$group
```

```{r}
table(lymphoid$group)
```

```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("Lymphoid_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("Lymphoid_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}

```

### HGG and LGG

```{r}
lymphoid$group <- gsub(" ","_",paste0(lymphoid$progress,"_",lymphoid$annotation_V2))
conditions <- unique(lymphoid$progress)[order(unique(lymphoid$progress))][1:2]
clusters <- c("0_activated_CD8", "1_activated_CD4", "Memory_B_1", "Memory_B_2", "Memory_B_3", "NKT", "NK", "Naive_T_cells")
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(lymphoid) <- lymphoid$group
```


```{r}
table(lymphoid$group)
```

```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("Lymphoid_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("Lymphoid_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}

```

## Myeloid:

```{r}
de_clusters <- function(current_cluster) {
  groups <- gsub(" ","_",paste0(conditions,"_",current_cluster))
  both_markers <- FindMarkers(myeloid, ident.1=groups[1], ident.2=groups[2], test.use="MAST", min.pct=0.00, logfc.threshold = 0) # change name of object 
  marker_genes <- row.names(both_markers)
  mutate(both_markers, gene = marker_genes, cluster = current_cluster)
}
```


```{r}
table(myeloid$annotation_V2)
```

### NT and LGG

MHC2_low_BAM, MHC2_high_BAM , Neutrophil,Slamf9_macrophage,cDC2, Classical_monocytes, Non-classical_monocytes

```{r}
myeloid$group <- gsub(" ","_",paste0(myeloid$progress,"_",myeloid$annotation_V2))
conditions <- unique(myeloid$progress)[order(unique(myeloid$progress))][2:3]
clusters <- c("MHC2_low_BAM", "MHC2_high_BAM" , "Neutrophil","Slamf9_macrophage","cDC2", "Classical_monocytes", "Non-classical_monocytes")
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(myeloid) <- myeloid$group
```


```{r}
table(myeloid$group)
```

```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("Myeloid_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("Myeloid_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}

```

### HGG and LGG

MHC2_low_BAM, MHC2_high_BAM , Neutrophil,Slamf9_macrophage,cDC2, Classical_monocytes, cDC1, pDC, Complement_GBP_BAM, intM,migDC

```{r}
myeloid$group <- gsub(" ","_",paste0(myeloid$progress,"_",myeloid$annotation_V2))
conditions <- unique(myeloid$progress)[order(unique(myeloid$progress))][1:2]
clusters <- c("MHC2_low_BAM", "MHC2_high_BAM" , "Neutrophil","Slamf9_macrophage","cDC2", "Classical_monocytes", "cDC1", "pDC", "Complement_GBP_BAM", "intM","migDC")
plts <- list()
padj_thres = 10e-2
log2FC_thres = 1
Idents(myeloid) <- myeloid$group
```


```{r}
table(myeloid$group)
```

```{r}
for (i in clusters) {
  output_file <- paste0(outdir,paste0("Myeloid_",conditions[1],'_vs_',conditions[2],'_', i, '.tsv'))
  output_file_flt <- paste0(outdir,paste0("Myeloid_",conditions[1],'_vs_',conditions[2],'_', i, '_filtered.tsv'))
  de_results_tbl <- lapply(i, de_clusters) 
  de_results_tbl <- de_results_tbl %>% bind_rows() %>%
			arrange(desc(abs(avg_log2FC))) %>%
			mutate(threshold = case_when(
			(p_val_adj < padj_thres) & (abs(avg_log2FC) > log2FC_thres) ~ "Pass",
			.default = "Fail"
			))
  write.table(de_results_tbl, output_file,sep="\t",row.names=T)
  
  de_results_tbl <- de_results_tbl %>% 
			filter(p_val_adj < padj_thres) %>%
			filter(abs(avg_log2FC) > log2FC_thres)
			
  write.table(de_results_tbl, output_file_flt,sep="\t",row.names=T)
}

```



