---
title: "visualization of scRNA-seq data"
output: html_document
date: "2024-07-30"
---

```{r}
library(Seurat)
library(tidyverse)

library(RColorBrewer) # for making palettes
library(gridExtra) # for arranging multiple plots
```

```{r}
pal <- c(brewer.pal(n = 12,name = "Set3"),brewer.pal(n = 12,name = "Paired"),brewer.pal(n = 8,name = "Accent"))
pal[2] <- "darkred"
pal[9] <- "grey25"
pal[23] <- "turquoise4"
pal[19] <- colors()[60]
pal[13] <- "darkblue"
pal[28] <- colors()[16]
pal[31] <- colors()[54] 
pal <- c(pal,colors()[c(85,84,83,89,45,31,41,42,26,29,10,139,107,108,120,109,119,121,143)])

```

# DimPlot and Feature Plot variants

## Split-by DimPlot with customized colors

```{r}
ann_cols <- pal[1:length(unique(Glioma$annotation_V2))]
anns <- unique(Glioma$annotation_V2)
names(ann_cols) <- anns[order(unique(Glioma$annotation_V2))]

plist <-  c("No tumor", "LGG","HGG")
grob <- lapply(plist, function(x){
  DimPlot(subset(Glioma, subset = progress == x), group.by = "annotation_V2",label=F,label.box = F,label.color = "white",repel = F)+scale_color_manual(values =  ann_cols,guide = "none")+scale_fill_manual(values = ann_cols,guide = "none")+guides(fill="none")+ggtitle(x)
})

g <- gridExtra::grid.arrange(grobs = grob, ncol = 3 )
ggsave("Glioma_V3_ann_V2_DimPlot_splitby_progress_unlabeled.pdf",plot = g,width = 20,height = 6)

```

## co-expression feature plot
```{r}
feature_mat <- as.data.frame(t(as.matrix(Glioma[["RNA"]]$data[c("Mki67","Top2a"),])))
feature_mat <- cbind(feature_mat,Glioma@reductions$umap@cell.embeddings)



cell_no_exp <- feature_mat %>% filter(Mki67 == 0 & Top2a == 0)

cell_exp <- feature_mat %>% filter(Mki67 > 0 & Top2a > 0) %>%
 mutate(
   Mki67_scale = percent_rank(Mki67), # code red
   Top2a_scale = - percent_rank(Top2a), # code blue
   co_exp_scale = Mki67_scale + Top2a_scale
)

ggplot() +
  geom_point(data = cell_no_exp,aes(x = umap_1, y = umap_2), color = "grey90",alpha = 0.3)+
  geom_point(data = cell_exp,aes(x = umap_1, y = umap_2,color = co_exp_scale),alpha = 0.5)+scale_color_gradientn(colours = c("blue2","red2"))+
  labs(color = "Mki67(red) Top2a(blue) co-expression: ")+theme_classic()
```


# proportion plots templates: details see specific Rmd

```{r}
g <- ggplot(cluster_sample_prop_table %>% filter(compartment == "neoplastic"))+
  geom_boxplot(aes(x = annotation_V1, y = prop_cluster_in_sample,fill = progress)) +
  geom_jitter(aes(x = annotation_V1, y = prop_cluster_in_sample,shape = progress), size=0.4, alpha=0.9) +
  RotatedAxis()
ggsave(filename = paste0(outdir,"neoplastic_cluster_sample_prop_boxplots_by_progress_ann_V1.pdf"),plot = g, width = 10,height = 10)

g <- ggplot(cluster_progress_prop %>% filter(compartment == "neoplastic")) +
  geom_bar(aes(x = progress, y = prop_cluster_in_progress,fill = annotation_V1),stat = "summary",position = "stack") +
  RotatedAxis()+
  scale_fill_manual(values = pal)+
  theme_classic()

g <- ggplot(cluster_sample_prop_table_batch_ID %>% filter(compartment == "neoplastic")) +
  geom_bar(aes(x = sample.id, y = prop_cluster_in_sample,fill = annotation_V1),stat = "summary",position = "stack") +
  facet_wrap(~progress, scales = "free")+
  RotatedAxis()+
  scale_fill_manual(values = pal)+
  theme_classic()
ggsave(filename = paste0(outdir,"neoplastic_cluster_sample_prop_barplots_by_progress_ann_V1.pdf"),plot = g, width = 8,height = 8)
```


# volcano plot

```{r}
HGG_LGG <- read.table(file = "DE_analysis_post_meeting2/Glioma_HGG_vs_LGG_tumor_5.tsv",header = T,sep = "\t")

g <- ggplot(HGG_LGG,aes(x=avg_log2FC, y = -log10(p_val_adj), label = gene))+
  geom_point(aes(color = threshold))+
  geom_vline(xintercept = c(1,-1),color = "red2")+geom_hline(yintercept = 1.3, color = "red2")+ ggrepel::geom_label_repel(aes(label=ifelse((p_val_adj < 0.05) & (abs(avg_log2FC) > 1),as.character(gene),'')),
                  box.padding   = 0.35, 
                  point.padding = 0.5,
                  size = 2,
                  segment.color = 'grey50',
                  max.overlaps = 30) +
  theme_classic()
g
ggsave("HGG_vs_LGG_tumor_5_volcano.pdf",plot = g, width = 10, height = 10)
```
