---
title: "Codes and RDS version overview"
output: html_document
date: "2024-08-06"
---

## Rmd and Robj versions:

**Archived prefix: Rmd involving "parallel universe"**
**Aux prefix: Rmd for dataset inspection but not formal analysis**
===============================pre-V1============================================

-- Archived_0-0-p (2 notebooks): 
--- cellranger raw alignment_out --> soupX --> compare with filtered alignment_out
--- conclusion: use filtered alignment output for data processing and analysis

===============================V1-processing============================================

-- 0-1, 0-5, Archived_0-5: 3 notebooks consolidated into *consolidated_1_prelim_processing.Rmd*
--- 0-1 (section 1): filtered align_out --> process single sample with miQC and obtained 5% estimation for mt.percent cutoff --> 5% mt.percent trimming --> **NT-C.Robj**, **T-C.Robj**, **NT-KO.Robj**, **T-KO.Robj** --> merge --> **whole.Robj**
--- 0-5_DoubletRemoval (section 2): **NT-C.Robj**, **T-C.Robj**, **NT-KO.Robj**, **T-KO.Robj** --> sample-wise DoubletFinder --> *Doublet removed version of each sample (name omitted)* --> merge --> **whole_DoubletFinder.Robj** --> select resolution 0.6 and pick clusters based on T cell markers -->  **T_cells_DF_res06.Robj**
--- Archived_0-5_Integrative_Analysis (section 3):**whole.Robj** --> **whole_processed_wHarmony.Robj** --> **whole_processed_wHarmony_miQC.Robj**
---- Archived_0-5-1_T_cell_processing:Archived version generating **whole_processed_wHarmony_miQC.Robj** --> **CD3E_high.Robj** and **CD3E_low.Robj**. Not used for analysis.

-- Aux_0-6: **whole_DoubletFinder.Robj** --> checking markers of major cell populations.

================================V1-analysis===========================================

-- Archived_1-0_T_cells_processing_DF_res06:  **T_cells_DF_res06.Robj** --> processing and select resolution=1 remove c18 which is Cd3e low --> **T_cells_DF_res06_V2.Robj**  ;**T_cells_DF_res06.Robj** --> TCR-seq analysis V1(more like walk-through) --> TCR QC --> **Cd8_TCR_filtered_V1.Robj** and **Cd4_TCR_filtered_V1.Robj**
--- version no longer active; mostly experimental

-- Archived_1-0-1_T_cell_subset: 
-- Archived_1-0-1_TCR_GEX_integrative: **whole_processed_wHarmony.Robj** + TCR out --> 

-- 1-0_T_cells_processing_DF_res06:  **T_cells_DF_res06.Robj** --> processing and select resolution=1 remove c18 which is Cd3e low --> **T_cells_DF_res06_V2.Robj** ; TCR-seq analysis + TCR QC --> **T_cells_DF_res06_V2_wClon.Robj** and **T_cells_DF_res06_wTCR.Robj**, **CD8_byTCR_V1.Robj**, **CD4_byTCR_V1.Robj**.

V1 conclusion: 5% mt.percent threshold + DF still has low quality cells

================================V2-processing========================================

Consider whole DF + 5% mt as V2, start with V2p1 by additional removal of cells marked as `miQC.keep == FALSE`

-- 2-0_processing_V2.Rmd: **whole_DoubletFinder.Robj** --> all V2 objects used for analysis were generated until artifact cluster were cleaned, resulting in V3 and V3p1.
--- QC finished whole dataset were divided into subsets for clustering-based QC and analysis. the final V2p1 whole was generated by merging back the subsets.
--- TCR-seq analysis in the same directory was deleted because codes were no longer used for the actual analysis and it was a duplicate from a previous TCR-seq analysis.

===============================V2-analysis============================================

2-4: scRep clonality analysis
2-5: cross-condition DE for conventional T cells
Aux 2-6: troubleshooting missing cells in SCN and weird cluster in V2 T cells.

=================================V3-finalization-and-publication-preparation=================
V3 are finalized versions for publication preparation
3_UPDATE: remove artifact cluster to generate V3 T cells

V3: odd cluster removal
V3p1: additional odd cluster removal

Pub1~Pub7: Notebooks in chronological orders.
V3p2: update in labels ??? one CD8 cluster contains NKT-like cells based on TCR and gene expression.
V3p3 late version of conventional T cells with ribo genes regressed out. 





